### PR TITLE
bugFix/epoch-start-miniblocks-pending-processed

### PR DESCRIPTION
**Pull request by sasurobert**
Recalibrated how processed and pending miniblocks are calculated in case of start in epoch. Added requests on processing of start in epoch to request the missing mteablocks.

### CODE REVIEW

File: process/block/shardblock.go [hunk 1]
```diff
@@ -276,15 +274,28 @@ func (sp *shardProcessor) requestEpochStartInfo(header *block.Header, haveTime f
 
 	go sp.requestHandler.RequestMetaHeader(header.EpochStartMetaHash)
 
+	headersPool := sp.dataPool.Headers()
 	for {
-		time.Sleep(time.Millisecond)
+		time.Sleep(100 * time.Millisecond)
```
Comment (iulianpascalau) [inline process/block/shardblock.go:279]: magic values

Comment (sasurobert) [reply process/block/shardblock.go:279]: done

File: process/block/shardblock.go [hunk 2]
```diff
@@ -276,15 +274,28 @@ func (sp *shardProcessor) requestEpochStartInfo(header *block.Header, haveTime f
...
 		if haveTime() < 0 {
 			break
 		}
 
 		if sp.epochStartTrigger.IsEpochStart() {
 			return nil
 		}
+
+		gotHdr, err := headersPool.GetHeaderByHash(header.EpochStartMetaHash)
```
Comment (SebastianMarian) [inline process/block/shardblock.go:290]: epochStartMetaBlock

File: process/block/shardblock.go [hunk 3]
```diff
@@ -276,15 +274,28 @@ func (sp *shardProcessor) requestEpochStartInfo(header *block.Header, haveTime f
...
+		if err != nil || check.IfNil(gotHdr) {
```
Comment (SebastianMarian) [inline process/block/shardblock.go:291]: err != nil would is enough

File: process/block/shardblock.go [hunk 4]
```diff
@@ -399,6 +411,21 @@ func (sp *shardProcessor) checkEpochCorrectness(
 	return nil
 }
 
+func (sp *shardProcessor) getOldEpochStartMetaHdr(header *block.Header) (data.HeaderHandler, error) {
```
Comment (SebastianMarian) [inline process/block/shardblock.go:417]: Unused

File: process/block/shardblock.go (hunk 2 above)
Comment (sasurobert) [reply process/block/shardblock.go:290]: done

File: process/block/shardblock.go (hunk 3 above)
Comment (sasurobert) [reply process/block/shardblock.go:291]: done

File: process/block/shardblock.go (hunk 4 above)
Comment (sasurobert) [reply process/block/shardblock.go:416]: done

File: process/block/shardblock.go [hunk 5]
```diff
@@ -399,6 +411,21 @@ func (sp *shardProcessor) checkEpochCorrectness(
...
+	metaBlock, err := sp.dataPool.Headers().GetHeaderByHash(header.EpochStartMetaHash)
```
Comment (SebastianMarian) [inline process/block/shardblock.go:417]: I think this line could be replaced with:
	metaBlock, err := process.GetMetaHeader(header.EpochStartMetaHash, sp.dataPool.Headers(), sp.marshalizer, sp.store)

In this way also header.EpochStartMetaHash would be searched first in pool and then if it is not found there in cacher->storage

Comment (sasurobert) [reply process/block/shardblock.go:417]: no need for that, as in case of storage it should be saved with the epoch trigger key.

Comment (SebastianMarian) [reply process/block/shardblock.go:417]: ok

File: epochStart/metachain/epochStartData_test.go [hunk 6]
```diff
@@ -378,3 +431,112 @@ func TestMetaProcessor_CreateEpochStartFromMetaBlockShouldWork(t *testing.T) {
 	err = epoch.VerifyEpochStartDataForMetablock(&block.MetaBlock{EpochStart: *epStart})
 	assert.Nil(t, err)
 }
+
+func TestMetaProcessor_CreateEpochStartFromMetaBlockEdgeCaseChecking(t *testing.T) {
```
Comment (iulianpascalau) [inline epochStart/metachain/epochStartData_test.go:435]: :+1:

**Pull request merged**